server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/CreateSnapshotRequest.java

```diff
@@ -364,6 +373,23 @@ public CreateSnapshotRequest userMetadata(@Nullable Map<String, Object> userMeta
         return this;
     }
 
+    /**
+     * Set a uuid to identify snapshot.
+     * If no uuid is specified, one will be created within SnapshotService
+     */
+    public CreateSnapshotRequest uuid(String uuid) {
+        this.uuid = uuid;
+        return this;
+    }
+
+    public String uuid() {
+        // If request was constructed on node with transport version < REGISTER_SLM_STATS it will not have a uuid yet
+        if (this.uuid == null) {
+            this.uuid = UUIDs.randomBase64UUID();
+        }
+        return this.uuid;
+    }
```

Comment:
I don't think there are multiple threads on the same node using this. Thus, though it would be easy to make threadsafe, I'm inclined to mark it as non-threadsafe so it doesn't give the impression that it needs to be threadsafe.

I've been thinking about this trying to come up with scenarios that could behave incorrectly. 
I'm not sure if this is really possible but I imagine a scenario where two uuid's might be generated for the same request.  Say there are 3 nodes with versions: 9, 7, 9.
- the request is constructed on a node with version 9 and thus has a uuid generated
- transferred through node with version 7, thus loses uuid
- lastly sent to the other node with version 9, which generates a new uuid.

Do you think it would be an option to fail requests in `validate()` if uuid is null? This would cause failure on snapshots requests during upgrades, but seems safer.

Reply:
I don't think there are multiple threads using this either. Marking it is fine. I also don't think that the 9->7->9 scenario is going to happen either, since the master node is the one doing these things, and not transferring it both and forth between nodes. So I think documenting it is a fine solution.

Reply:
I am a bit confused about `master node is the one doing these things,`. In the case of SnapshotLifecycleTask that makes sense I think as SnapshotLifecycleService is only running on the master node. But for ad-hoc snapshot created through the [rest api](https://github.com/elastic/elasticsearch/blob/5662f988b208718b8b3a1f1a1b0e549ac0a2d563/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCreateSnapshotAction.java#L46) is this true? (Although I agree my example seems pretty darn contrived)

Reply:
I think you are right that this is fine as is, I'm just wanting to understand if something like this is possible.

Also, I'm thinking I'll split this getter into two methods: a getter `uuid()` and `getOrCreateUUID()` as making a "getter" not thread-safe seems to defy expectations.